### PR TITLE
PartitionPredicate support for map operations

### LINKED ISSUE #11962: IMap aggregate, projection, localKeySet &  executeOnEntries methods throws exception when PartitionPredicate used
See: https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java#L797

A Partition predicate check can be added & query can be sent only that partition, This will also improve the performance of the aggregations. See `Map.values` with Predicate as an example: https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java#L658

### PR DESCRIPTION
**Pull request by gokhanoner**
Fixes #11962

Added `PartitionPredicate` support to `localKeySet`, `aggregate`, `projection` & `executeOnEntries` methods on member side IMap implementation.

### CODE REVIEW

**Comment by gokhanoner** [comment]
run-lab-run

File: hazelcast/src/test/java/com/hazelcast/map/PartitionPredicateTest.java [hunk 1]
```diff
@@ -111,6 +128,43 @@ public void entries() {
         }
     }
 
+    @Test
```
Comment (mmedenjak) [inline hazelcast/src/test/java/com/hazelcast/map/PartitionPredicateTest.java]: Can you add the same tests on the `ClientPartitionPredicateTest` as well?

File: hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java [hunk 2]
```diff
@@ -767,79 +733,37 @@ public ICompletableFuture submitToKey(K key, EntryProcessor entryProcessor) {
...
-        AggregationResult result = queryEngine.execute(query, Target.ALL_NODES);
-        return result.<R>getAggregator().aggregate();
+        return aggregate(aggregator, TruePredicate.<K, V>truePredicate());
     }
 
     @Override
     public <R> R aggregate(Aggregator<Map.Entry<K, V>, R> aggregator, Predicate<K, V> predicate) {
         checkNotNull(aggregator, NULL_AGGREGATOR_IS_NOT_ALLOWED);
         checkNotNull(predicate, NULL_PREDICATE_IS_NOT_ALLOWED);
         checkNotPagingPredicate(predicate, "aggregate");
```
[the remark this answers is not in the document]
Comment (gokhanoner) [reply hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxyImpl.java]: @mmedenjak it precludes `PagindPredicate`, not `PartitionPredicate`. See `checkNotPagingPredicate` method.

File: hazelcast/src/test/java/com/hazelcast/map/PartitionPredicateTest.java (hunk 1 above)
Comment (gokhanoner) [reply hazelcast/src/test/java/com/hazelcast/map/PartitionPredicateTest.java]: @mmedenjak `ClientPartitionPredicateTest ` contains similar tests for `aggregate`, `project` and  `executeOnEntries` already. @sancar fixed the client side in 3.9.2. See https://github.com/hazelcast/hazelcast/pull/11961

**Comment by mmedenjak** [comment]
Looks good. It may be worth noting that the `PartitionPredicates` will only work if you upgrade all of your members to 3.9.3 and that it might won't work:
- when running a mixed 3.9.0, 3.9.1, 3.9.2 and 3.9.3 cluster
- when rolling up from 3.8.x to 3.9.x

**Comment by gokhanoner** [comment]
@mmedenjak we can ask @Serdaro to put that info to release notes?

**Comment by mmedenjak** [comment]
@gokhanoner sure. @Serdaro can you put this in the release notes - https://github.com/hazelcast/hazelcast/pull/12158#issuecomment-358568027

**Comment by Serdaro** [comment]
👍

**Pull request merged**